lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,572 @@
+//!
+//! TODO: Fill this in once the API is finalized.
+//!
+//! # Note
+//!
+//! The [`Route`] is compute before each payment attempt. Any updates to the [`NetworkGraph`] or
+//! changes to channels scores should be applied prior to retries. This typically means any
+//! [`EventHandler`] decorators responsible for this should decorate [`PaymentRetryHandler`] so that
+//! such changes take effect before retrying.
+
+use crate::Invoice;
+
+use bitcoin::hashes::Hash;
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::secp256k1::key::PublicKey;
+
+use lightning::ln::{PaymentHash, PaymentSecret};
+use lightning::ln::channelmanager::{ChannelDetails, PaymentSendFailure};
+use lightning::ln::features::InvoiceFeatures;
+use lightning::ln::msgs::LightningError;
+use lightning::routing::network_graph::NetworkGraph;
+use lightning::routing::router::{Route, RouteHint};
+use lightning::util::events::{Event, EventHandler};
+use lightning::util::logger::Logger;
+
+use std::cell::RefCell;
+use std::collections::hash_map::{self, HashMap};
+use std::ops::Deref;
+use std::sync::Mutex;
+
+/// The maximum number of payment attempts before forwarding the [`Event::PaymentFailed`] to be
+/// handled.
+const MAX_PAYMENT_ATTEMPTS: usize = 3;
+
+/// A utility for paying [`Invoice]`s.
+pub struct InvoicePayer<P: Deref, R, G>
+where P::Target: Payer, R: Router, G: Deref<Target = NetworkGraph> {
+	payer: P,
+	router: R,
+	network_graph: G,
+	invoice_cache: Mutex<HashMap<PaymentHash, Invoice>>,
+}
+
+/// A trait defining behavior of an [`Invoice`] payer.
+pub trait Payer {
+	/// Returns the payer's node id.
+	fn node_id(&self) -> PublicKey;
+
+	/// Returns the payer's channels.
+	fn first_hops(&self) -> Vec<ChannelDetails>;
+
+	/// Sends a payment over the Lightning Network using the given [`Route`].
+	fn send_payment(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+	) -> Result<(), PaymentSendFailure>;
+}
+
+/// A trait defining behavior for routing an [`Invoice`] payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` using the provided first and last hops and
+	/// [`NetworkGraph`] for a payment with the given values.
+	fn find_route(
+		&self,
+		payer: &PublicKey,
+		network: &NetworkGraph,
+		payee: &PublicKey,
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+/// An [`EventHandler`] decorator for retrying failed payments.
+pub struct PaymentRetryHandler<I, P: Deref, R, G, L: Deref, E>
```

Comment:
Separate from `InvoicePayer`, right? `Payer` is the trait parameter.

As of now, the ownership model is such that any decorating `EventHandler` (or `BackgroundProcessor` for that matter) would take ownership of this, which means `InvoicePayer` could then only be used as an event handler and no longer for initiating payments. For instance, we'll need to decorate this with `NetworkUpdateHandler` so that updates are applied before retries.

I suppose we could implement `EventHandler` for `T: Deref<Target = InvoicePayer<...>>` and users could use an `Arc<InvoicePayer<...>>` (haven't actually tried this). But it seems cleaner to separate paying from the retry behavior, which also lends to intuitive naming.